server/app_sudoku/use_cases/post_solve.py

```diff
@@ -0,0 +1,15 @@
+from typing import List
+
+from server.app_sudoku.domain.sudoku import Sudoku
+
+
+class InvalidSudoku(Exception):
+    pass
+
+
+def solve_sudoku(sudoku_board: List[int]) -> List[int]:
+    sudoku = Sudoku(sudoku_board)
+
+    if not sudoku.is_valid:
+        raise InvalidSudoku("Not a valid sudoku")
+    return sudoku.solved
```

Comment:
Jag måste erkänna att jag inte riktigt förstår varför detta behövs. Här importeras Sudoku och i `sudoku_controller` importeras `solve_sudoku`. Men skulle inte Sudoku-klassen kunna importeras direkt i `sudoku_controller` och vi definierar detta direkt i den filen? Eller är det för att kunna definiera flera klasser med exceptions? Onödiga lager av abstraktioner kan ju göra koden svårare att läsa, eller vad tror du @taleldayekh?

Reply:
Återkommer snart med tanken bakom detta 🙂.

Reply:
Denna fil hör till `service layer` som är lagret mellan `business logic layer` där Sudoku klassen ligger och `interface layer` där vi har `sudoku_controller.py`.

Idén med denna fil (och lager) är att den ska orkestrera allt som är nödvändigt för ett “use case”.  I detta fal är vårt use case att lösa en sudoku. För tillfället är allt den gör är att instantiera Sudoku klassen och försöka lösa sudokun. Men det går att tänka sig ett scenario där vi exempelvis utöver att lösa en sudoku också vill hålla räkning över hur många gånger en användare har valt att få en sudoku löst, om vår app skulle vara ett spel exempelvis, och ta med det i beräkningen när vi delar ur ett “score”.

I ovanstående exempel, utöver att returnera en löst sudoku så skulle vi kanske också returnera hur många gånger användaren begär ut en lösning.

Som du ser har vi inte bara Sudoku klassen att förhålla oss till längre utan kanske också en Scores klass eller något liknande. I denna fil skulle då dessa behöva komma samman och orkestrera/göra det som behövs för att returnera relevant data till interface lagret.

Det är detta som vi separerar ut från interface lagret genom denna fil. På så sätt bekymrar sig `sudoku_controller.py` bara om att returnera korrekt `Response` baserat på sitt use case. I och med det kan vi enkelt byta ut Flask mot något annat ramverk och allt vi behöver göra är att plugga in det nya ramverket med respektive endpoint till sitt respektive use case.

För tillfället sker inte mycket och det kan därför framstå som onödig abstraktion men det skapar en flexibel kodbas som ifall den växer kan göra det på ett sätt där alla komponenter är `decoupled`.

Så har jag förstått meningen med sådan arkitektur 🙂.

Reply:
Grymt! Tack för utförligt svar, nu förstår jag hur det hänger ihop! Det påminner mig om att ofta när man kollar på objektorienterade programmeringsexempel då där är en ganska liten kod verkar allt så abstrakt och onödigt krångligt. Men sedan när man skalar upp koden blir sådana abstraktioner och uppdelningar nödvändiga. Vårt projekt är än så länge litet att det verkade som ett onödigt lager men nu förstår jag hur det hänger ihop och kan förstå att när man skalar blir detta också nödvändigt :